### ISSUE TITLE
Optimise hits and scatter recording, and switch to gen. xy for uncertainties in track fitting

### ISSUE THREAD

**Issue by GilesStrong** [GilesStrong/tomopt#105]
The recorded hits saved by `DetectorPanel` and `VoxelDetectorLayer` depend on the resolution of the panel, however the `_get_hit_uncs` function of `VoxelScatterBatch` and `PanelScatterBatch` recomputes the resolution and efficiency of the hits based on the reco position.
The ideal approach would be to save the resolution and efficiency in the hit information, however this resulted in errors when computing Jacobians, due to in-place modification of the tensor and I was unable to find out what was causing this. I've since found that it is due to the xy position of the muon changing as the muon propagates, and when the reco hit is produced, the noise depends on the resolution which is computed using a "reference" to the muon xy, rather than a copy of the xy position. Since the muon position are modified in-place, the xy position changes between the hit being recorded, and the Jacobian of the hits w.r.t the resolution. This can be solved by using `mu.xy.detach().clone()` or `mu.xy.data` to explicitly base the resolution calculation on a copy of the xy position.
Making this change will reduce the need to compute the resolution and efficiency twice.

Additionally, the uncertainty used in fitting the tracks is based on the resolution at the recorded position, rather than the true position. In training mode with the Gaussian distributed resolution, this means that the uncertainty is not computed correctly. In realistic validation mode, hits are constrained to fall within the panel, if the muon hits the panel, and the resolution is uniform. Based on this the uncertainties should really be computed using the true positions of the muons.

Saving both the resolution and efficiency as part of the hit info will solve both of these problems.
